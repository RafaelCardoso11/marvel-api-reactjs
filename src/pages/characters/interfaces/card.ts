export interface ICard {
  id: number;
  name: string;
  thumbnail: { path: string; extension: string };
}
