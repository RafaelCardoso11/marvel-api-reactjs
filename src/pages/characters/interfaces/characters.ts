export interface ICharacters {
  id: number;
  name: string;
  thumbnail: { path: string; extension: string };
}
