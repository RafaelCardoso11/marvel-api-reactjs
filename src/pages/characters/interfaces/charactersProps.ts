import { ICharacters } from "./characters";

export interface ICharactersProps {
   characters: ICharacters,
   index: number,
  }
  