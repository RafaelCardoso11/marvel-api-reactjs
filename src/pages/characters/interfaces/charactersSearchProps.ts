import { ICharacters } from "./characters";

export interface ICardsSearchProps {
  data: {
    results: ICharacters[];
  };
}
