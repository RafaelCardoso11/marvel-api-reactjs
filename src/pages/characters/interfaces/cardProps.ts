import { ICard } from "./card";

export interface ICardProps {
   card: ICard,
   index: number,
  }
  