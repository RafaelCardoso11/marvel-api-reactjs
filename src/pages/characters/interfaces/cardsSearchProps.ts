import { ICard } from "./card";

export interface ICardsSearchProps {
  data: {
    results: ICard[];
  };
}
