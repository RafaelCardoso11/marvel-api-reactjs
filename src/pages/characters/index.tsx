import { useDataFetcher } from "@/hooks/useDataFetcher";
import { useState } from "react";

import { ICardsSearchProps } from "./interfaces/charactersSearchProps";
import { Cards } from "./components/cards";
import { StyledContainer, StyledContainerSearch } from "./styles";

export const Characters = () => {
  const [queryParams, setQueryParams] = useState("characters");
  const [searchCharacters, setSearchCharacters] = useState<string>("");
  const { data: characters } = useDataFetcher<ICardsSearchProps>(queryParams);

  const handleSearch = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const searchCharacters = event.target.value;
    setSearchCharacters(searchCharacters);
    if (searchCharacters) {
      setQueryParams("characters?nameStartsWith=" + searchCharacters);
    } else {
      setQueryParams("characters");
    }
  };

  return (
    <StyledContainer data-testid="id-container-characters">
      <StyledContainerSearch data-testid="id-container-search-characters">
        <input
          type="text"
          placeholder="Pesquisar..."
          data-testid="id-searchCharacters"
          value={searchCharacters}
          onChange={handleSearch}
        />
      </StyledContainerSearch>
      {characters && <Cards data={characters.data} data-testid="id-cards-characters"/>}
    </StyledContainer>
  );
};
