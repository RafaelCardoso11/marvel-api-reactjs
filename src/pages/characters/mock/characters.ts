import { ICardsSearchProps } from "../interfaces/charactersSearchProps";

export const characters: ICardsSearchProps = {
  data: {
    results: [
      {
        id: 1011334,
        name: "3-D Man",
        thumbnail: {
          path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          extension: "jpg",
        },
      },
    ],
  },
};
