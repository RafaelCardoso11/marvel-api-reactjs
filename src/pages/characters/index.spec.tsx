import { Characters } from "./index";
import { cleanup, fireEvent, render, renderHook } from "@testing-library/react";
import { charactersMock } from "@/mocks/characters";
import { ICardsSearchProps } from "./interfaces/charactersSearchProps";
import * as useDataFetcherModule from "@/hooks/useDataFetcher";
import RoutersContextTesting from "@/tests/routers";

const charactersMockEmpty: ICardsSearchProps = {
  data: {
    results: [],
  },
};

const mockData = jest.fn();

jest.mock("@/hooks/useDataFetcher", () => {
  return {
    useDataFetcher: () => {
      return mockData();
    },
  };
});

describe("<Characters/>", () => {
  afterEach(cleanup);
  afterAll(() => {
    jest.unmock("@/hooks/useDataFetcher");
  });
  beforeEach(() => {
    // Resetar o mock antes de cada teste
    mockData.mockRestore();
    jest.clearAllMocks();
  });
  it("Deveria renderizar na tela os componentes container-characters, container-search-characters e searchCharacters", () => {
    mockData.mockReturnValue({
      data: charactersMockEmpty,
    });
    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );
    const containerCharacters = screen.getByTestId("id-container-characters");
    const containerSearchCharacters = screen.getByTestId(
      "id-container-search-characters"
    );
    const searchCharacters = screen.getByTestId("id-search-characters");

    expect(containerCharacters).toBeInTheDocument();
    expect(containerSearchCharacters).toBeInTheDocument();
    expect(searchCharacters).toBeInTheDocument();

    expect(screen.container).toMatchSnapshot();
    expect.assertions(4);
  });
  it("Deveria buscar os characters ao renderizar a aplicação", async () => {
    mockData.mockReturnValue({
      data: charactersMock,
    });
    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const useDataFetcherMock = jest.spyOn(
      useDataFetcherModule,
      "useDataFetcher"
    );

    const { result } = renderHook(() =>
      useDataFetcherModule.useDataFetcher("")
    );

    expect(useDataFetcherMock).toHaveBeenCalled();
    expect(result.current.data).toEqual(charactersMock);
    expect(screen.container).toMatchSnapshot();

    expect.assertions(3);
  });
  it("Deveria buscar os characters e listalos em cards na tela caso tenha characters", async () => {
    mockData.mockReturnValue({
      data: charactersMock,
    });

    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const { result } = renderHook(() =>
      useDataFetcherModule.useDataFetcher("")
    );

    expect(result.current.data).toEqual(charactersMock);

    const containerCardsCharacters = await screen.findByTestId(
      "id-cards-characters"
    );

    const imgsInScreen = screen.getAllByRole("img");

    expect(imgsInScreen.length).toBe(charactersMock.data.results.length);
    expect(containerCardsCharacters).toBeInTheDocument();
    expect(screen.container).toMatchSnapshot();
    expect.assertions(4);
  });
  it("Deveria buscar a partir do nome dos personagens", async () => {
    mockData.mockReturnValue({
      data: charactersMock,
    });

    const valueNameSearch = "Hulk";

    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const inputSearchCharacters = screen.getByTestId("id-search-characters");

    expect(inputSearchCharacters).toBeInTheDocument();

    fireEvent.input(inputSearchCharacters, {
      target: {
        value: valueNameSearch,
      },
    });
    expect(inputSearchCharacters).toHaveValue(valueNameSearch);

    const { result, rerender } = renderHook(() =>
      useDataFetcherModule.useDataFetcher<ICardsSearchProps>("")
    );

    const filterCharacterFromName = result.current.data.data.results.filter(
      ({ name }) =>
        name.toLocaleLowerCase().includes(valueNameSearch.toLocaleLowerCase())
    );
    const dataWithCharactersSearch = {
      data: {
        results: filterCharacterFromName,
      },
    };

    mockData.mockReturnValue({
      data: dataWithCharactersSearch,
    });

    screen.rerender(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );
    rerender();

    const cardHulk = screen.getByText("Hulk");
    const imgsInScreen = screen.getAllByRole("img");

    expect(cardHulk).toBeInTheDocument();
    expect(imgsInScreen.length).toBe(1);

    expect(screen.container).toMatchSnapshot();
    expect.assertions(5);
  });
  it("Deveria buscar a partir da inicial do nome dos personagens", () => {
    mockData.mockReturnValue({
      data: charactersMock,
    });

    const valueNameSearch = "H";

    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const inputSearchCharacters = screen.getByTestId("id-search-characters");

    expect(inputSearchCharacters).toBeInTheDocument();

    fireEvent.input(inputSearchCharacters, {
      target: {
        value: valueNameSearch,
      },
    });
    expect(inputSearchCharacters).toHaveValue(valueNameSearch);

    const { result, rerender } = renderHook(() =>
      useDataFetcherModule.useDataFetcher<ICardsSearchProps>("")
    );

    const filterCharacterFromName = result.current.data.data.results.filter(
      ({ name }) =>
        name.toLocaleLowerCase().includes(valueNameSearch.toLocaleLowerCase())
    );

    const dataWithCharactersSearch = {
      data: {
        results: filterCharacterFromName,
      },
    };

    mockData.mockReturnValue({
      data: dataWithCharactersSearch,
    });

    rerender();
    screen.rerender(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const cardHulk = screen.getByText("Hulk");
    const imgsInScreen = screen.getAllByRole("img");

    expect(cardHulk).toBeInTheDocument();
    expect(imgsInScreen.length).toBe(1);
    expect(screen.container).toMatchSnapshot();
    expect.assertions(5);
  });
  it("Deveria buscar todos os personagens caso a busca seja vazia", () => {
    mockData.mockReturnValue({
      data: charactersMockEmpty,
    });

    const valueNameSearch = " ";

    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const inputSearchCharacters = screen.getByTestId("id-search-characters");

    expect(inputSearchCharacters).toBeInTheDocument();

    fireEvent.change(inputSearchCharacters, {
      target: {
        value: valueNameSearch,
      },
    });
    expect(inputSearchCharacters).toHaveValue(valueNameSearch);

    const { result, rerender } = renderHook(() =>
      useDataFetcherModule.useDataFetcher<ICardsSearchProps>("")
    );

    mockData.mockReturnValue({
      data: charactersMock,
    });

    screen.rerender(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );
    rerender();

    const cardHulk = screen.getByText("Hulk");
    const imgsInScreen = screen.getAllByRole("img");

    expect(result.current.data).toEqual(charactersMock);
    expect(cardHulk).toBeInTheDocument();
    expect(imgsInScreen.length).toBe(charactersMock.data.results.length);
    expect(screen.container).toMatchSnapshot();
    expect.assertions(6);
  });
  it("Deveria ao buscar um characters que não existe mostrar uma mensagem que não foi encontrado", () => {
    mockData.mockReturnValue({
      data: charactersMock,
    });

    const valueNameSearch = "batman";

    const screen = render(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );

    const inputSearchCharacters = screen.getByTestId("id-search-characters");

    expect(inputSearchCharacters).toBeInTheDocument();

    fireEvent.change(inputSearchCharacters, {
      target: {
        value: valueNameSearch,
      },
    });
    expect(inputSearchCharacters).toHaveValue(valueNameSearch);

    const { result, rerender } = renderHook(() =>
      useDataFetcherModule.useDataFetcher<ICardsSearchProps>("")
    );

    mockData.mockReturnValue({
      data: charactersMockEmpty,
    });

    screen.rerender(
      <RoutersContextTesting>
        <Characters />
      </RoutersContextTesting>
    );
    rerender();

    const messageSearchNotFind = screen.getByText(
      /Nenhum personagem encontrado/i
    );

    expect(result.current.data).toEqual(charactersMockEmpty);
    expect(messageSearchNotFind).toBeInTheDocument();
    expect(screen.container).toMatchSnapshot();
    expect.assertions(5);
  });
});
