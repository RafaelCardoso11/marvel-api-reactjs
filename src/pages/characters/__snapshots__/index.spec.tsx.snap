// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Characters/> Deveria ao buscar um characters que não existe mostrar uma mensagem que não foi encontrado 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value="batman"
      />
    </div>
    Nenhum personagem encontrado
  </div>
</div>
`;

exports[`<Characters/> Deveria buscar a partir da inicial do nome dos personagens 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value="H"
      />
    </div>
    <div
      data-testid="id-cards-characters"
    >
      <section
        class="sc-gZGWnS ixooVO"
        data-testid="id-cards-container"
      >
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              Hulk
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<Characters/> Deveria buscar a partir do nome dos personagens 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value="Hulk"
      />
    </div>
    <div
      data-testid="id-cards-characters"
    >
      <section
        class="sc-gZGWnS ixooVO"
        data-testid="id-cards-container"
      >
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              Hulk
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<Characters/> Deveria buscar os characters ao renderizar a aplicação 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value=""
      />
    </div>
    <div
      data-testid="id-cards-characters"
    >
      <section
        class="sc-gZGWnS ixooVO"
        data-testid="id-cards-container"
      >
        <div
          class="sc-kvAqsT crfTLG"
        >
          <h2
            class="sc-cdtKKh hnyVce"
          >
            3-D Man
          </h2>
          <img
            alt=""
            class="sc-grJzXs hvxsKK"
            src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
          />
        </div>
        <div
          class="sc-kvAqsT crfTLG"
        >
          <h2
            class="sc-cdtKKh hnyVce"
          >
            Hulk
          </h2>
          <img
            alt=""
            class="sc-grJzXs hvxsKK"
            src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
          />
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<Characters/> Deveria buscar os characters e listalos em cards na tela caso tenha characters 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value=""
      />
    </div>
    <div
      data-testid="id-cards-characters"
    >
      <section
        class="sc-gZGWnS ixooVO"
        data-testid="id-cards-container"
      >
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              3-D Man
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              Hulk
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<Characters/> Deveria buscar todos os personagens caso a busca seja vazia 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value=" "
      />
    </div>
    <div
      data-testid="id-cards-characters"
    >
      <section
        class="sc-gZGWnS ixooVO"
        data-testid="id-cards-container"
      >
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              3-D Man
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
        <div
          data-testid="id-cards-content"
        >
          <div
            class="sc-kvAqsT crfTLG"
          >
            <h2
              class="sc-cdtKKh hnyVce"
            >
              Hulk
            </h2>
            <img
              alt=""
              class="sc-grJzXs hvxsKK"
              src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
            />
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`<Characters/> Deveria renderizar na tela os componentes container-characters, container-search-characters e searchCharacters 1`] = `
<div>
  <div
    class="sc-cLrhaH jGMCgM"
    data-testid="id-container-characters"
  >
    <div
      class="sc-jWWaec iQvkjV"
      data-testid="id-container-search-characters"
    >
      <input
        data-testid="id-search-characters"
        placeholder="Pesquisar..."
        type="text"
        value=""
      />
    </div>
    Nenhum personagem encontrado
  </div>
</div>
`;
